### PR TITLE
Instruction refactoring

### PR DESCRIPTION
**Pull request by avshvetsov**

### CODE REVIEW

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        } else if (second != null) {
+            stringBuilder.append(" " + second);
+        } else if (third != null) {
+            stringBuilder.append(" " + third);
+        } else if (fourth != null) {
+            stringBuilder.append(" " + fourth);
+        }
+        return stringBuilder.toString();
+    }
+}
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: please clean the code, give meaningful names

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        if (first != null) {
+            stringBuilder.append(" " + first);
+        } else if (second != null) {
+            stringBuilder.append(" " + second);
+        } else if (third != null) {
+            stringBuilder.append(" " + third);
+        } else if (fourth != null) {
+            stringBuilder.append(" " + fourth);
+        }
+        return stringBuilder.toString();
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: can be suppressed into the List of strings joining by a space

File: src/main/java/org/bytecodeparser/instruction/Instruction.java [hunk 3]
```diff
@@ -2,11 +2,24 @@
...
-public interface Instruction {
+public class Instruction {
+    final InstructionArguments args;
+    private final InstructionTypes type;
+
+    public Instruction(Integer opCode, DataInputStream dataInputStream) {
+        this.type = Arrays.stream(InstructionTypes.values())
+                .filter(instructionTypes -> opCode.equals(instructionTypes.opCode))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Unable to read instruction with opcode = " + opCode));
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/Instruction.java]: everytime Instruction is created - there is new stream created - it is inefficient - you can factor out it to the static Map where Key is opcode and the value is InstructionType instance

File: src/main/java/org/bytecodeparser/instruction/InstructionTypes.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    PUTFIELD("PUTFIELD", 181, simpleShort),
+    INVOKESPECIAL("INVOKESPECIAL", 183, simpleShort),
+    INVOKESTATIC("INVOKESTATIC", 184, simpleShort),
+    INVOKEINTERFACE("INVOKEINTERFACE", 185, tripleShortByteByte),
+    INVOKEDYNAMIC("INVOKEDYNAMIC", 186, tripleShortByteByte),
+    NEW("NEW", 187, simpleShort),
+    CHECKCAST("CHECKCAST", 192, simpleShort),
+    ;
+
+    final String name;
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionTypes.java]: there is no need in name I believe

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.bytecodeparser.utility.ThrowingFunction;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.function.Function;
+
+public enum PrettyPrintTypes {
+    simple(dataInputStream -> InstructionArguments.builder().build()),
+    simpleByte(resolveStrategyWrapper(dataInputStream -> InstructionArguments.builder().first(dataInputStream.readByte()).build())),
+    simpleShort(resolveStrategyWrapper(dataInputStream -> InstructionArguments.builder().first((dataInputStream.readByte() << 8) | dataInputStream.readByte()).build())),
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: make constants as per conventions - with big letters

File: src/test/java/org/bytecodeparser/instruction/InstructionTest.java [hunk 6]
```diff
@@ -14,27 +14,27 @@
...
+//
+//  @ParameterizedTest
+//  @MethodSource("getInstructions")
+//  void allInstructionsShouldHaveDataInputStreamConstructor(Class<? extends Instruction> inst) {
+//    assertThat(inst.getDeclaredConstructors())
+//        .withFailMessage(() -> inst.getSimpleName() + " instruction doesn't contain constructor with DataInputStream")
+//        .anyMatch(constructor -> Arrays.stream(constructor.getParameterTypes()).collect(Collectors.toSet())
+//            .equals(Set.of(DataInputStream.class)))
+//    ;
+//  }
```
Comment (Dany41) [inline src/test/java/org/bytecodeparser/instruction/InstructionTest.java]: delete or rework

**Comment by avshvetsov** [comment]
@Dany41 merge pls
Tests was deleted, cause I have no inspiration)) Will be added next time.

File: src/main/java/org/bytecodeparser/instruction/Instruction.java [hunk 7]
```diff
@@ -2,11 +2,29 @@
...
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
 @CustomPrettyPrint
-public interface Instruction {
+public class Instruction {
+
+    private static final Map<Integer, InstructionTypes> instructions = Arrays.stream(InstructionTypes.values())
+            .collect(Collectors.toMap(instructionTypes -> instructionTypes.opCode, Function.identity()));
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/Instruction.java]: create getter in instruction types and use method referwence

File: src/main/java/org/bytecodeparser/instruction/Instruction.java [hunk 8]
```diff
@@ -2,11 +2,29 @@
...
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
 @CustomPrettyPrint
-public interface Instruction {
+public class Instruction {
+
+    private static final Map<Integer, InstructionTypes> instructions = Arrays.stream(InstructionTypes.values())
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/Instruction.java]: EnumSet is designed to be efficient, it is best practice to use it, not values()

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package org.bytecodeparser.instruction;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class InstructionArguments {
+
+    private InstructionTypes type;
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: InstructionArguments is included in Instruction instance. InstructionTypes is included here, in InstructionArguments , and in Instruction. There is no need in such circularity, you can do either invoking ``` type.prettyPrintTypes.prettyPrintArgumentsStrategy.apply(args) ``` in Instruction prettyPrint or make prettyPrint generic in instructionArguments as it was

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package org.bytecodeparser.instruction;
+
+import org.bytecodeparser.utility.ThrowingFunction;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.function.Function;
+
+public enum PrettyPrintTypes {
+    SIMPLE(dataInputStream -> InstructionArguments.builder().build(), arguments -> ""),
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: name 'SIMPLE' is not informative, you can use EMPTY

Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: for every SIMPLE instruction you use InstructionArguments.builder().build(). Every time allocating object but there is no need in this - factor out it in the static singleton instance and reuse everytime

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package org.bytecodeparser.instruction;
+
+import org.bytecodeparser.utility.ThrowingFunction;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.function.Function;
+
+public enum PrettyPrintTypes {
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java:9]: why class with name PrettyPrintTypes do something with DataInoutStream?

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+    static Function<DataInputStream, InstructionArguments> resolveStrategyWrapper(ThrowingFunction<DataInputStream, IOException, InstructionArguments> function) {
+        return dataInputStream -> {
+            try {
+                return function.apply(dataInputStream);
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+        };
+    }
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java:46]: move to ThrowingFunction interface

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+    private byte indexByte;
+    private short indexShort;
+    private byte count;
+    private byte zero;
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: there can be multiple zeros

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package org.bytecodeparser.instruction;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class InstructionArguments {
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionArguments.java:8]: you have finite number of PrettyPrintTypes, so you can create finite corresponding number of InstructionArguments implementations

File: src/main/java/org/bytecodeparser/instruction/InstructionTypes.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    IDIV(108, SIMPLE),
+    IRETURN(172, SIMPLE),
+    ARETURN(176, SIMPLE),
+    RETURN(177, SIMPLE),
+    GETFIELD(180, SIMPLE_SHORT),
+    PUTFIELD(181, SIMPLE_SHORT),
+    INVOKESPECIAL(183, SIMPLE_SHORT),
+    INVOKESTATIC(184, SIMPLE_SHORT),
+    INVOKEINTERFACE(185, TRIPLE_SHORT_BYTE_BYTE),
+    INVOKEDYNAMIC(186, TRIPLE_SHORT_BYTE_BYTE),
```
Comment (Dany41) [inline src/main/java/org/bytecodeparser/instruction/InstructionTypes.java]: TRIPLE_SHORT_BYTE_BYTE is wrong for this instruction

File: src/main/java/org/bytecodeparser/instruction/Instruction.java (hunk 8 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/Instruction.java]: It is not agree with your point: https://stackoverflow.com/questions/2464950/enum-values-vs-enumset-allof-which-one-is-more-preferable

Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/Instruction.java]: But thx! before I don't know about EnumSet)

Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/Instruction.java]: values is method which is generated by compiler and aims for optimiozation (you will not find it in source code, but will find reflective calls), EnumSet.allOf() uses it under the hood. There are no doubts in performance, it is about the conventional way how to work with collections of enum constants

Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/Instruction.java]: ok, +

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 9 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: Fixed. Choose first option, cause don't understand idea of second

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 10 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: fixed

Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: this was done to unify all Function<DataInputStream, InstructionArguments> resolveStrategy with each other
unify prefer efficiency in this case. IMHO.

But I try to implement your idea and it looks bad. Check it.

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 11 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java:9]: PrettyPrintTypes describes the logic of working with DataInoutStream and o nothing with it directly.

... or I don't get your point...

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 12 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java:46]: I have tried do it, but have no idea how to do it

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 13 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: I haven't seen any examples of it. 
My proposition: when someone will be add this case to InstructionTypes, he should resolve "multiple zeros" problem in InstructionArguments.

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 14 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java:8]: do you propose to create InstructionArguments classes for each PrettyPrintTypes?

File: src/main/java/org/bytecodeparser/instruction/InstructionTypes.java (hunk 15 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/InstructionTypes.java]: image
why?

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 13 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: Do you understand purpose of zero operand at all? Could you explain it?

File: src/main/java/org/bytecodeparser/instruction/InstructionTypes.java (hunk 15 above)
Comment (avshvetsov) [reply src/main/java/org/bytecodeparser/instruction/InstructionTypes.java]: * fix for INVOKEDYNAMIC

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 14 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java:8]: yeap

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 11 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java:9]: I mean it is confusing naming. PrettyPrintTypes supposed to describe different types of printing somethin pretty, but it also does dataInputStream manipulation. Maybe smth like InstructionArgumentResolver

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 10 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: move it to InstructionArguments class

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 13 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java]: it is used by InvokeDynamic - in runtime this instruction is substituted with another one, maybe tha's why there are extra bytes

File: src/main/java/org/bytecodeparser/instruction/InstructionArguments.java (hunk 14 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/InstructionArguments.java:8]: but if we do it we can delete PrettyPrintTypes.. ok, lets leave it as it is now

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 11 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java:9]: will do it with follow up commit by myself

File: src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java (hunk 10 above)
Comment (Dany41) [reply src/main/java/org/bytecodeparser/instruction/PrettyPrintTypes.java]: will do it with follow up commit by myself

**Review by Dany41 — APPROVED** [review]
thanks for your work!

**Pull request merged**